src/lib/chart/dendrogram.typ

```diff
@@ -0,0 +1,312 @@
+///
+///                 *Example*
+///                 ```typc
+///                 ((1, 2, 0.5), (3, 4, 1), (6, 7, 2), (5, 8, 2.5))
+///                 ```
+/// - x1-key (int,string): Key to access the first cluster of a data row. This 
+///                        key is used as argument to the rows `.at(..)` 
+///                        function.
+/// - x2-key (int,string): Key to access the second cluster of a data row. 
+///                        This key is used as argument to the rows `.at(..)` 
+///                        function.
+/// - value-key (int,string): Key(s) to access value(s) of data row.
+///                           These keys are used as argument to the
+///                           rows `.at(..)` function.
+/// - mode (string): Chart mode:
+///                  - `"vertical"` -- Vertically displayed dendrogram
+/// - size (array): Chart size as width and height tuple in canvas units;
+///                 height can be set to `auto`.
+/// - line-style (style,function): Style or function (idx => style) to use for
+///                               each leaf, accepts a palette function.
+/// - x-label (content,none): X axis label
+/// - y-tick-step (float): Step size of y axis ticks 
+/// - x-ticks (array): List of tick values or value/label tuples
+///
+///                    *Example*
+///                    
+///                    `(1, 5, 10)` or `((1, [One]), (2, [Two]), (10, [Ten]))`
+/// - y-ticks (array): List of tick values or value/label tuples
+///
+///                    *Example*
+///                    
+///                    `(1, 5, 10)` or `((1, [One]), (2, [Two]), (10, [Ten]))`
+/// - y-unit (content,auto): Tick suffix added to each tick label
+/// - y-label (content,none): Y axis label
+/// - y-min (number,auto): Y axis minimum value
+/// - y-max (number,auto): Y axis maximum value
+#let dendrogram(data,
+                x1-key: 0,
+                x2-key: 1,
+                height-key: 2,
+                size: (auto, 1),
+                mode: "vertical",
+                line-style: (stroke: black + 1pt),
+                x-label: none,
+                y-tick-step: auto,
+                x-ticks: auto,
+                y-ticks: (),
+                y-unit: auto,
+                y-label: none,
+                y-min: auto,
+                y-max: auto,
+                ) = {
+  import draw: *
+
+  assert(mode in dendrogram-modes,
+    message: "Invalid dendrogram mode. Use: " + repr(dendrogram-modes))
+  assert(type(x1-key) in (int, str))
+  assert(type(x2-key) in (int, str))
+  assert(type(height-key) in (int, str))
+
+  if size.at(0) == auto {
+    size.at(0) = (data.len() + 2)
+  }
+
+  let max-value = (dendrogram-max-value-fn.at(mode))(data, height-key)
+  if y-max != auto {
+    max-value = y-max
+  }
+  let min-value = (dendrogram-min-value-fn.at(mode))(data, height-key)
+  if y-min != auto {
+    min-value = y-min
+  }
+
+  let x-ticks = x-ticks
```

Comment:
You shouldn't need to do this. It'll only panic if you try and modify an external variable from inside a closure/function. But as this is passed as a parameter it'll be fine.

Reply:
done